packages/components/src/settings-form/index.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * External dependencies
+ */
+import { Button } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { Form } from '../index';
+import {
+	SettingText,
+	SettingPassword,
+	SettingCheckbox,
+	SettingSelect,
+} from './types';
+
+const typeMap = {
+	text: SettingText,
+	password: SettingPassword,
+	checkbox: SettingCheckbox,
+	select: SettingSelect,
+	default: SettingText,
+};
+
+export const SettingsForm = ( {
+	fields: baseFields = [],
+	isBusy,
+	onSubmit = () => {},
+	onButtonClick = () => {},
+	onChange = () => {},
+	validate = () => {},
+	buttonLabel,
+} ) => {
+	// Support accepting fields in the format provided by the API (object), but transform to Array
+	const fields =
+		baseFields instanceof Array ? baseFields : Object.values( baseFields );
+
+	const getInitialConfigValues = () => {
+		if ( fields ) {
+			return fields.reduce(
+				( data, field ) => ( {
+					...data,
+					[ field.id ]: field.value,
+				} ),
+				{}
+			);
+		}
+	};
+
+	return (
+		<Form
+			initialValues={ getInitialConfigValues() }
+			onChangeCallback={ onChange }
+			onSubmitCallback={ onSubmit }
+			validate={ validate }
+		>
+			{ ( { getInputProps, handleSubmit } ) => {
+				return (
+					<div className="woocommerce-component-settings">
+						{ fields.map( ( field ) => {
+							if ( field.type && ! ( field.type in typeMap ) ) {
+								/* eslint-disable no-console */
+								console.warn(
+									`Field type of ${ field.type } not current supported in SettingsForm component`
+								);
+								/* eslint-enable no-console */
+								return null;
+							}
+
+							const Control = typeMap[ field.type || 'default' ];
```

Comment:
Sorry for changing up here, but this logic may need some tweaking.  There's a possibility that this is defined by doesn't match one of the types (e.g., `textarea`) in which case I think we shouldn't attempt to show the control at all if we don't have a matching component.

Reply:
Did you see the block of code right above this @joshuatf ? That's its intention. If a non-supported type is provided it won't render it at all, and will print a warning to the console.